include/dca/phys/dca_step/cluster_solver/ctint/walker/ctint_walker_cpu_submatrix.hpp

```diff
@@ -0,0 +1,974 @@
+  else
+    nbr_of_steps_ = nb_steps_per_sweep_;
+
+  // Get the maximum of Monte Carlo steps/moves that can be performed during one submatrix step.
+  max_nbr_of_moves = parameters_.getMaxSubmatrixSize();
+
+  BaseClass::n_steps_ += nbr_of_steps_;
+  while (nbr_of_steps_ > 0) {
+    nbr_of_moves_to_delay_ = std::min(nbr_of_steps_, max_nbr_of_moves);
+    nbr_of_steps_ -= nbr_of_moves_to_delay_;
+
+    doStep();
+  }
+
+  BaseClass::updateSweepAverages();
+}
+
+template <class Parameters, typename Real>
+void CtintWalkerSubmatrix<linalg::CPU, Parameters, Real>::doStep() {
+  generateDelayedMoves(nbr_of_moves_to_delay_);
+  doSubmatrixUpdate();
+}
+
+// Do one step with arbitrary number of moves. For testing.
+template <class Parameters, typename Real>
+void CtintWalkerSubmatrix<linalg::CPU, Parameters, Real>::doStep(const int nbr_of_movesto_delay) {
+  std::cout << "\nStarted doStep() function for testing." << std::endl;
+
+  generateDelayedMoves(nbr_of_movesto_delay);
+
+  std::cout << "\nGenerated " << nbr_of_movesto_delay << " moves for testing.\n" << std::endl;
+
+  doSubmatrixUpdate();
+}
+
+template <class Parameters, typename Real>
+void CtintWalkerSubmatrix<linalg::CPU, Parameters, Real>::generateDelayedMoves(
+    const int nbr_of_movesto_delay) {
+  assert(nbr_of_movesto_delay > 0);
+
+  delayed_moves_.clear();
+
+  int nbr_of_moves = 0;
+
+  for (int s = 0; s < 2; ++s) {
+    n_init_[s] = configuration_.getSector(s).size();
+  }
+  n_ = config_size_init_ = configuration_.size();
+
+  // Generate delayed moves.
+  while (nbr_of_moves < nbr_of_movesto_delay) {
+    DelayedMoveType delayed_move;
+
+    delayed_move.move_type = generateMoveType();
+
+    switch (delayed_move.move_type) {
+      case REMOVAL:
+        delayed_move.removal_rng = rng_();
+        break;
+
+      case INSERTION:
+        configuration_.insertRandom(rng_);
+        for (int i = 0; i < configuration_.lastInsertionSize(); ++i)
+          delayed_move.indices.push_back(configuration_.size() -
+                                         configuration_.lastInsertionSize() + i);
+        break;
+
+      default:
+        throw(std::logic_error("Unknown move type encountered."));
+    }
+
+    delayed_move.acceptance_rng = rng_();
+    delayed_moves_.push_back(delayed_move);
+
+    ++nbr_of_moves;
+  }
+
+  for (int s = 0; s < 2; ++s) {
+    n_max_[s] = configuration_.getSector(s).size();
+  }
+
+  const int config_size_final = configuration_.size();
+  conf_removal_list_.resize(config_size_final - config_size_init_);
+  std::iota(conf_removal_list_.begin(), conf_removal_list_.end(), config_size_init_);
+}
+
+template <class Parameters, typename Real>
+void CtintWalkerSubmatrix<linalg::CPU, Parameters, Real>::doSubmatrixUpdate() {
+  computeMInit();
+  computeGInit();
+  mainSubmatrixProcess();
+  updateM();
+}
+
+template <class Parameters, typename Real>
+void CtintWalkerSubmatrix<linalg::CPU, Parameters, Real>::mainSubmatrixProcess() {
```

Comment:
I think is way too long. It would be much easier to understand what happens here if this was refactored into a series of well named function/method calls.

Reply:
I will need to think on this one.